### PR TITLE
Migrated level IDs to strings

### PR DESCRIPTION
**Pull request by jmkulwik**
Note: This PR was a shared effort between jmkulwik and cforkish.

this is the third and final installment of work to make `id` a required string property on `level` objects in our client code. initially we decided to use numbers for the id property, and that work was completed in #38085. then after being educated about the preference for using strings for keys in javascript objects (which is the primary purpose of the id prop), we decided to update the prop to be a string. this work was divided into two parts. the first was to update all the levelbuilder code to use strings, which was completed in #38403. this PR completes the work by updating the rest of our client code to use strings.

### CODE REVIEW

File: apps/src/code-studio/components/progress/LessonProgress.jsx [hunk 1]
```diff
@@ -83,7 +83,9 @@ class LessonProgress extends Component {
     lessonTrophyEnabled: PropTypes.bool,
     width: PropTypes.number,
     setDesiredWidth: PropTypes.func,
-    currentPageNumber: PropTypes.number
+    // currentPageNumber is intentionally a string since it gets used
+    // as a string in the URL
+    currentPageNumber: PropTypes.string
```
Comment (maureensturgeon) [inline apps/src/code-studio/components/progress/LessonProgress.jsx]: It looks like this gets compared to another number https://github.com/code-dot-org/code-dot-org/blob/f4a46c17366ffa4c08eb937e05bd1aacddc9893b/apps/src/code-studio/components/progress/LessonProgress.jsx#L209 (unless that's also a string) Seems like it would make more sense to leave this as a number especially since the variable name has Number in it

File: apps/src/sites/studio/pages/levels/_bubble_choice.js [hunk 2]
```diff
@@ -5,7 +5,14 @@ import BubbleChoice from '@cdo/apps/code-studio/components/BubbleChoice';
 const script = document.querySelector('script[data-bubblechoice]');
 const data = JSON.parse(script.dataset.bubblechoice);
 
+let level = data.level;
```
Comment (maureensturgeon) [inline apps/src/sites/studio/pages/levels/_bubble_choice.js:8]: nit `const`

File: apps/src/templates/progress/progressHelpers.js [hunk 3]
```diff
@@ -181,7 +184,7 @@ export const processedLevel = level => {
     levelNumber: level.kind === LevelKind.unplugged ? undefined : level.title,
     isConceptLevel: level.is_concept_level,
     bonus: level.bonus,
-    pageNumber: level.page_number,
+    pageNumber: level.page_number && level.page_number.toString(),
```
Comment (maureensturgeon) [inline apps/src/templates/progress/progressHelpers.js]: I don't think this will work if `page_number` is 0, is that a case we should consider? (not sure if we'd ever end up with a 0 page number..)

File: apps/test/unit/code-studio/components/BonusLevelsTest.jsx [hunk 4]
```diff
@@ -11,7 +11,7 @@ const DEFAULT_PROPS = {
 };
 
 describe('BonusLevels', () => {
-  it('renders correct number of sublevels', () => {
+  it('renders correct number of bonus levels', () => {
```
Comment (maureensturgeon) [inline apps/test/unit/code-studio/components/BonusLevelsTest.jsx:14]: ✨

File: dashboard/app/views/layouts/_header.html.haml [hunk 5]
```diff
@@ -102,8 +102,7 @@
       lesson_extras_enabled = view_as.lesson_extras_enabled?(script)
     end
 
-    # don't trust outside content in parameter :puzzle_page - should be integer, so immediately call to_i
-    puzzle_page = params[:puzzle_page] ? params[:puzzle_page].to_i : ApplicationHelper::PUZZLE_PAGE_NONE
+    puzzle_page = params[:puzzle_page] ? params[:puzzle_page] : ApplicationHelper::PUZZLE_PAGE_NONE
```
Comment (maureensturgeon) [inline dashboard/app/views/layouts/_header.html.haml:105]: Is `ApplicationHelper::PUZZLE_PAGE_NONE` a string?

**Review by maureensturgeon — APPROVED** [review]

File: apps/src/code-studio/components/progress/LessonProgress.jsx (hunk 1 above)
Comment (jmkulwik) [reply apps/src/code-studio/components/progress/LessonProgress.jsx]: This is a good question. The page number is conceptually similar to a sub-level ID. This was why I decided to swap it to a string. That being said, I don't think we ever use it in a way that would care whether it's a number or a string. So it very well may make sense to just change it back to being a number.

**Review by jmkulwik — COMMENTED** [review]
Looks like I can't sign off on a PR I opened, so count this as my ✔️ 
LGTM. I'd recommend a short bug bash as I recall there being a couple tricky spots, but I'm fine with whatever you decide. Thank you for taking this across the finish line!

EDIT: ❌ I do have 2 required changes for a true sign-off 
1. Update the description to be a more thorough explanation of what is happening in this PR
2. Merge it directly into staging separate from the level-id-required branch (this will require merging level-id-required into staging first)

File: apps/src/code-studio/components/progress/LessonProgress.jsx (hunk 1 above)
Comment (cforkish) [reply apps/src/code-studio/components/progress/LessonProgress.jsx]: yeah i think that makes sense. nice catch @maureensturgeon!

File: apps/src/templates/progress/progressHelpers.js (hunk 3 above)
Comment (cforkish) [reply apps/src/templates/progress/progressHelpers.js]: good catch. i will check for `undefined` to be safe.

File: dashboard/app/views/layouts/_header.html.haml (hunk 5 above)
Comment (cforkish) [reply dashboard/app/views/layouts/_header.html.haml:105]: no its an int, so this line doesn't need to change.

**Comment by jmkulwik** [comment]
> ~EDIT: ❌ I do have 2 required changes for a true sign-off~
> 
> 1. Update the description to be a more thorough explanation of what is happening in this PR
> 2. Merge it directly into staging separate from the level-id-required branch (this will require merging level-id-required into staging first)

These two are all good now. Taking a second quick look through the code.

File: apps/src/code-studio/components/progress/LessonProgress.jsx [hunk 6]
```diff
@@ -203,11 +203,7 @@ class LessonProgress extends Component {
             {lessonTrophyEnabled && <div style={styles.spacer} />}
             {levels.map((level, index) => {
               let isCurrent = level.isCurrentLevel;
-              if (
-                isCurrent &&
-                level.kind === LevelKind.assessment &&
-                level.pageNumber
-              ) {
+              if (isCurrent && level.kind === LevelKind.assessment) {
```
Comment (jmkulwik) [inline apps/src/code-studio/components/progress/LessonProgress.jsx:206]: Is this change intentional?

Comment (jmkulwik) [reply apps/src/code-studio/components/progress/LessonProgress.jsx:206]: (may need to rebase on staging)

Comment (cforkish) [reply apps/src/code-studio/components/progress/LessonProgress.jsx:206]: yeah this is intentional. i had thought i was responsible for fixing the single-page assessment issue, so in my branch i had set `PUZZLE_PAGE_NONE` as the default page number rather than leaving it undefined, which makes this check work for single page assessments because `currentPageNumber` and `level.pageNumber` are both `PUZZLE_PAGE_NONE`.
